src/net/gosha/atproto/firehose.clj

```diff
@@ -0,0 +1,91 @@
+(ns net.gosha.atproto.firehose
+  (:require
+   [clojure.core.async    :as async]
+   [clojure.tools.logging :as log]
+   [charred.api           :as json])
+  (:import
+   [java.net URI]
+   [org.java_websocket.client WebSocketClient]))
+
+(def parse-fn 
+  (json/parse-json-fn 
+   {:key-fn  keyword
+    :profile :mutable ; Use mutable datastructures for better performance
+    :async?  false    ; Disable async for small messages
+    :bufsize 8192}))  ; Smaller buffer size for small messages}))
+
+(defn parse-message
+  [msg]
+  (try
+    (let [data (parse-fn msg)]
+      (case (:kind data)
+        "commit" 
+        {:type       :commit
+         :did        (:did data)
+         :timestamp  (:time_us data)
+         :rev        (get-in data [:commit :rev])
+         :operation  (get-in data [:commit :operation])
+         :collection (get-in data [:commit :collection])
+         :record     (get-in data [:commit :record])
+         :raw        data}
+        
+        {:type :unknown
+         :data data}))
+    (catch Exception e
+      (log/error "Parse error:" (.getMessage e))
+      nil)))
+
+(defn create-websocket-client
+  [uri output-ch]
+  (doto 
+    (proxy [WebSocketClient] [(URI. (str uri "?wantedCollections=app.bsky.feed.post"))]
+      (onOpen [_]
+        (log/info "Connected to firehose"))
+      
+      (onClose [code reason remote]
+        (log/info "Disconnected from firehose:" reason))
+      
+      (onMessage [message]
+        (when-let [parsed (parse-message message)]
+          (async/>!! output-ch parsed)))
```

Comment:
This will block the thread if the channel's buffer is full and it isn't consumed fast enough.

What is the behavior of WebSocketClient in this case? We should be intentional about this. (empirically: the endpoint times out since it can't respond to pings if its threads are all blocked). 

Ideally we could specify a large-ish buffer so the system could handle being backed up by ~30 seconds (e.g, if a downstream service crashed and reboots). 

But we should also log a warning if the buffer gets large, and then start dropping messages if the buffer is full rather than blocking the websocket thread (blocking a live processing thread is typically a bad idea unless you know exactly what you're doing)

Reply:
Ok, so this is the one I have been stuck on. I think I have refactored it to allow what you are asking but here is the issue. Your hunch was correct and we are overrunning our buffer. My log ` Buffer full - dropping message. Consider increasing buffer size or processing messages faster.` is getting printed rather quickly. I'm not sure how to fix or account for this.